### PR TITLE
chore(snap): refactor code to use blocking queues with threads for SNAP messages processing

### PR DESCRIPTION
**Pull request by Vovchyk**
## Description

changes:
- client SNAP sync state migrated from `SnapshotProcessor` to `SnapSyncState`
- processing of SNAP requests and responses is being queued in appropriate blocking queue and then executed in a dedicated thread 
- now there are two separate snap configs for client and server
- updated a bit snap integration test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Requires Activation Code (Hard Fork)

* **Other information**:

### CODE REVIEW

**Review by fmacleal — APPROVED** [review]
Very good job. You did an amazing and important refactor improving one of the most sensitive issues we had on the snapshot synch. Congrats.

I just have a few recommendations of refactor, really no big deal found. 🙂

File: rskj-core/src/test/java/co/rsk/net/sync/SyncMessageHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        //when
+        jobQueue.put(new SyncMessageHandler.Job(mock(Peer.class), mock(Message.class)) {
+            @Override
+            public void run() {
+                jobCalled.set(true);
+            }
+        });
+
+        //then
+        thread.join(THREAD_JOIN_TIMEOUT);
```
Comment (nagarev) [inline rskj-core/src/test/java/co/rsk/net/sync/SyncMessageHandlerTest.java:77]: Mini idea: I would move this `thread.join(...)` to the `//when` block as it's not part of the "check" step.

Comment (nagarev) [reply rskj-core/src/test/java/co/rsk/net/sync/SyncMessageHandlerTest.java:77]: Same [here](https://github.com/rsksmart/rskj/pull/2500/files#diff-dc006ec78d028f9a5c118de85bf07d9eda8cefb44e46bb9273e795ef99573484R116) and [here](https://github.com/rsksmart/rskj/pull/2500/files#diff-dc006ec78d028f9a5c118de85bf07d9eda8cefb44e46bb9273e795ef99573484R138).

**Review by nagarev — COMMENTED** [review]
LGTM, nice refactorings! Left a couple of minor comments, let's see the discussions on fred open comments

File: rskj-core/src/main/java/co/rsk/net/messages/MessageVisitor.java
[the remark this answers is not in the document]
Comment (Vovchyk) [reply rskj-core/src/main/java/co/rsk/net/messages/MessageVisitor.java:1]: `syncProcessor` is sorta "stateful" and preserves active state (`private SyncState syncState;` field). On the other hand, `snapshotProcessor` class doesn't manage it's own state, just depends on others

so basically we have this chain of deps: `MessageVisitor -> SyncProcessor -> SnapSyncState -> SnapshotProcessor` for responses, as we need to hold state in `SnapSyncState`,

.. and `MessageVisitor -> SnapshotProcessor` for requests, as we do not preserve client sessions

**Review by asoto-iov — APPROVED** [review]

File: rskj-core/src/integrationTest/java/co/rsk/snapshotsync/SnapshotSyncIntegrationTest.java [hunk 2]
```diff
@@ -89,33 +90,37 @@ public void whenStartTheServerAndClientNodes_thenTheClientWillSynchWithServer()
         serverNode.startNode();
         ThreadTimerHelper.waitForSeconds(30);
 
+        JsonNode serverBestBlockResponse = OkHttpClientTestFixture.getJsonResponseForGetBestBlockMessage(portServerRpc, "latest");
+        String serverBestBlockNumber = serverBestBlockResponse.get(0).get("result").get("number").asText();
```
Comment (jurajpiar) [inline rskj-core/src/integrationTest/java/co/rsk/snapshotsync/SnapshotSyncIntegrationTest.java:94]: In theory, any of these `.get(0).get("result").get("number")` could fail, right?

**Review by fmacleal — APPROVED** [review]
Good job, I liked of the refactor and the result in general. This refactor is tricky and touch in an important and sensitive port. Well done.
👌 

I was wondering if maybe we could try to create an integration test with multiple requests to a server to test some concurrency. Of course this could be added in a feature task, this one here is already too long.

**Pull request merged**